src/runtime/test/client_negotiation_test.cpp

```diff
@@ -0,0 +1,114 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "runtime/security/negotiation_utils.h"
+#include "runtime/security/client_negotiation.h"
+#include "runtime/rpc/network.sim.h"
+
+#include <gtest/gtest.h>
+#include <dsn/utility/flags.h>
+
+namespace dsn {
+namespace security {
+class client_negotiation_test : public testing::Test
+{
+public:
+    client_negotiation_test()
+    {
+        std::unique_ptr<tools::sim_network_provider> sim_net(
+            new tools::sim_network_provider(nullptr, nullptr));
+        _sim_session = sim_net->create_client_session(rpc_address("localhost", 10086));
+        _client_negotiation = new client_negotiation(_sim_session);
+    }
+
+    void on_recv_mechanism(const negotiation_response &resp)
+    {
+        _client_negotiation->on_recv_mechanisms(resp);
+    }
+
+    void handle_response(error_code err, const negotiation_response &resp)
+    {
+        _client_negotiation->handle_response(err, std::move(resp));
+    }
+
+    const std::string &get_selected_mechanism() { return _client_negotiation->_selected_mechanism; }
+
+    negotiation_status::type get_negotiation_status() { return _client_negotiation->_status; }
+
+    // _sim_session is used for holding the sim_rpc_session which is created in ctor,
+    // in case it is released. Because negotiation keeps only a raw pointer.
+    rpc_session_ptr _sim_session;
+    client_negotiation *_client_negotiation;
+};
+
+TEST_F(client_negotiation_test, on_recv_mechanisms)
+{
+    struct
+    {
+        negotiation_status::type resp_status;
+        std::string resp_msg;
+        std::string selected_mechanism;
+    } tests[] = {{negotiation_status::type::SASL_SELECT_MECHANISMS, "GSSAPI", ""},
+                 {negotiation_status::type::SASL_LIST_MECHANISMS_RESP, "TEST", ""},
+                 {negotiation_status::type::SASL_LIST_MECHANISMS_RESP, "GSSAPI", "GSSAPI"}};
```

Comment:
Add test case to test response with multiple mechanisms

Reply:
good idea.